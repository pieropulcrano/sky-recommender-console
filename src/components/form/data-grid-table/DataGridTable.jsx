import React from 'react';
import PropTypes from 'prop-types';
import { useField, useFormikContext } from 'formik';
import { DataGridTableWrapper, DataTable } from './DataGridTable.styled';
import { ErrorMsg } from '../error-msg/ErrorMsg.styled';

/**
 * Table (connected with Formik state by the param "name") to display data.
 */

const MIN_ROWS = 25;
const MAX_ROWS = 50;

const DataGridTable = ({ name, rows, columns, ...props }) => {
  const [selectionModel, setSelectionModel] = React.useState([]);
  const [pageSize, setPageSize] = React.useState(MIN_ROWS);
  const { setFieldValue } = useFormikContext();
  const [, meta] = useField(name);

  const handlePageSizeChange = (newPageSize) => setPageSize(newPageSize);

  const handleSelectionModelChange = (selection) => {
    if (selection.length > 1) {
      const selectionSet = new Set(selectionModel);
      const result = selection.filter((s) => !selectionSet.has(s));
      setSelectionModel(result);
      setFieldValue(name, rows.filter((row) => row.id === result[0])[0]);
    } else {
      setSelectionModel(selection);
      setFieldValue(name, rows.filter((row) => row.id === selection[0])[0]);
    }
  };

  return (
    <DataGridTableWrapper>
      <DataTable
        rows={rows}
        columns={columns}
        pageSize={pageSize}
        rowsPerPageOptions={[MIN_ROWS, MAX_ROWS]}
        checkboxSelection
        hideFooterSelectedRowCount
        onPageSizeChange={handlePageSizeChange}
        selectionModel={selectionModel}
        onSelectionModelChange={handleSelectionModelChange}
        renderAsync={true}
        {...props}
        sx={{
          '& .MuiDataGrid-columnHeaders': {
            borderTopLeftRadius: '0px',
            borderTopRightRadius: '0px',
          },
        }}
      />
      {meta.error && <ErrorMsg>Pick a value from the result list.</ErrorMsg>}
    </DataGridTableWrapper>
  );
};

DataGridTable.propTypes = {
  /**
   * The field name that connects the table with the form state.
   */
  name: PropTypes.string.isRequired,
  /**
   * The data displayed by the table.
   */
  rows: PropTypes.array.isRequired,
  /**
   * The configuration of all the columns of the table.
   */
  columns: PropTypes.array.isRequired,
};

export default DataGridTable;
