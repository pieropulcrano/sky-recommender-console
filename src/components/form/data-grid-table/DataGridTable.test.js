import React from 'react';
import { Formik } from 'formik';
import * as Yup from 'yup';
import { render, screen, waitFor, fireEvent } from '@testing-library/react';
import DataGridTable from './DataGridTable';

describe('DataGridTable', () => {
  const props = {
    name: 'data-table',
    rows: [
      { id: '1', field_1: 'field-1', field_2: 'field_2', field_3: 'field_3' },
    ],
    columns: [
      { field: 'field_1', headerName: 'field-1', width: '150' },
      { field: 'field_2', headerName: 'field-2', width: '150' },
      { field: 'field_3', headerName: 'field-3', width: '150' },
    ],
  };

  const initialValues = {
    selectedEvent: null,
  };

  const validationSchema = Yup.object().shape({
    selectedEvent: Yup.object().required(),
  });

  it('should render', () => {
    render(
      <Formik initialValues={initialValues} validationSchema={validationSchema}>
        <DataGridTable {...props} />
      </Formik>,
    );

    expect(document.body.childNodes[0].children).toMatchSnapshot();
  });

  it("should allow user to select a value from the result's list", async () => {
    render(
      <Formik initialValues={initialValues} validationSchema={validationSchema}>
        <DataGridTable {...props} />
      </Formik>,
    );

    const rowCheckBox = screen.queryByLabelText('Select row');
    await waitFor(() => {
      fireEvent.click(rowCheckBox);
    });

    expect(rowCheckBox.checked).toEqual(true);

    await waitFor(() => {
      fireEvent.click(rowCheckBox);
    });

    expect(rowCheckBox.checked).toEqual(false);
  });
});
