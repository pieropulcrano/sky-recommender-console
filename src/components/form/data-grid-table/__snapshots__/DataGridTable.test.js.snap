// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataGridTable should render 1`] = `
HTMLCollection [
  <div
    class="css-x8ilf6"
  >
    <div
      aria-colcount="4"
      aria-multiselectable="false"
      aria-rowcount="1"
      class="MuiDataGrid-root css-j1c84w-MuiDataGrid-root"
      role="grid"
    >
      <div>
        <div />
      </div>
      <div
        class="MuiDataGrid-main css-204u17-MuiDataGrid-main"
      >
        <div
          class="MuiDataGrid-columnHeaders css-f3jnds-MuiDataGrid-columnHeaders"
          style="min-height: 56px; max-height: 56px; line-height: 56px;"
        >
          <div
            aria-rowindex="1"
            class="MuiDataGrid-columnHeadersInner MuiDataGrid-columnHeadersInner--scrollable css-1s0hp0k-MuiDataGrid-columnHeadersInner"
            role="row"
            style="transform: translate3d(0px, 0px, 0px);"
          >
            <div
              aria-colindex="1"
              class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--alignCenter MuiDataGrid-columnHeaderCheckbox"
              data-field="__check__"
              role="columnheader"
              style="width: 50px; min-width: 50px; max-width: 50px;"
              tabindex="0"
            >
              <div
                class="MuiDataGrid-columnHeaderDraggableContainer"
                draggable="false"
              >
                <div
                  class="MuiDataGrid-columnHeaderTitleContainer"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainerContent"
                  >
                    <span
                      class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiDataGrid-checkboxInput css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                    >
                      <input
                        aria-label="Select all rows"
                        class="PrivateSwitchBase-input css-1m9pwf3"
                        data-indeterminate="false"
                        tabindex="-1"
                        type="checkbox"
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CheckBoxOutlineBlankIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                style="min-height: 56px; opacity: 1;"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                  data-testid="SeparatorIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 19V5h2v14z"
                  />
                </svg>
              </div>
            </div>
            <div
              aria-colindex="2"
              class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--sortable"
              data-field="field_1"
              role="columnheader"
              style="width: 150px; min-width: 150px; max-width: 150px;"
              tabindex="-1"
            >
              <div
                class="MuiDataGrid-columnHeaderDraggableContainer"
                draggable="false"
              >
                <div
                  class="MuiDataGrid-columnHeaderTitleContainer"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainerContent"
                  >
                    <div
                      aria-label=""
                      class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                      data-mui-internal-clone-element="true"
                    >
                      field-1
                    </div>
                  </div>
                  <div
                    class="MuiDataGrid-iconButtonContainer css-ltf0zy-MuiDataGrid-iconButtonContainer"
                  >
                    <button
                      aria-label="Sort"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                      tabindex="-1"
                      title="Sort"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-sortIcon css-ptiqhd-MuiSvgIcon-root"
                        data-testid="ArrowUpwardIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
                <div
                  class="MuiDataGrid-menuIcon"
                >
                  <button
                    aria-controls="mui-6"
                    aria-haspopup="true"
                    aria-label="Menu"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiDataGrid-menuIconButton css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                    id="mui-7"
                    tabindex="-1"
                    title="Menu"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                      data-testid="TripleDotsVerticalIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
              <div
                class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                style="min-height: 56px; opacity: 1;"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                  data-testid="SeparatorIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 19V5h2v14z"
                  />
                </svg>
              </div>
            </div>
            <div
              aria-colindex="3"
              class="MuiDataGrid-columnHeader MuiDataGrid-columnHeader--sortable"
              data-field="field_2"
              role="columnheader"
              style="width: 150px; min-width: 150px; max-width: 150px;"
              tabindex="-1"
            >
              <div
                class="MuiDataGrid-columnHeaderDraggableContainer"
                draggable="false"
              >
                <div
                  class="MuiDataGrid-columnHeaderTitleContainer"
                >
                  <div
                    class="MuiDataGrid-columnHeaderTitleContainerContent"
                  >
                    <div
                      aria-label=""
                      class="MuiDataGrid-columnHeaderTitle css-1jbbcbn-MuiDataGrid-columnHeaderTitle"
                      data-mui-internal-clone-element="true"
                    >
                      field-2
                    </div>
                  </div>
                  <div
                    class="MuiDataGrid-iconButtonContainer css-ltf0zy-MuiDataGrid-iconButtonContainer"
                  >
                    <button
                      aria-label="Sort"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                      tabindex="-1"
                      title="Sort"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall MuiDataGrid-sortIcon css-ptiqhd-MuiSvgIcon-root"
                        data-testid="ArrowUpwardIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M4 12l1.41 1.41L11 7.83V20h2V7.83l5.58 5.59L20 12l-8-8-8 8z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                </div>
                <div
                  class="MuiDataGrid-menuIcon"
                >
                  <button
                    aria-controls="mui-9"
                    aria-haspopup="true"
                    aria-label="Menu"
                    class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall MuiDataGrid-menuIconButton css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
                    id="mui-10"
                    tabindex="-1"
                    title="Menu"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
                      data-testid="TripleDotsVerticalIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                      />
                    </svg>
                    <span
                      class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                    />
                  </button>
                </div>
              </div>
              <div
                class="MuiDataGrid-columnSeparator MuiDataGrid-columnSeparator--sideRight"
                style="min-height: 56px; opacity: 1;"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiDataGrid-iconSeparator css-i4bv87-MuiSvgIcon-root"
                  data-testid="SeparatorIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M11 19V5h2v14z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <div
          style="overflow: visible; height: 0px; width: 0px;"
        >
          <div
            class="MuiDataGrid-virtualScroller css-1pans1z-MuiDataGrid-virtualScroller"
            style="width: 0px; height: auto; margin-top: 56px; overflow-x: hidden;"
          >
            <div
              class="MuiDataGrid-virtualScrollerContent MuiDataGrid-virtualScrollerContent--overflowed css-jz7yqw-MuiDataGrid-virtualScrollerContent"
              style="width: auto; height: 52px; min-height: auto;"
            >
              <div
                class="MuiDataGrid-virtualScrollerRenderZone css-s1v7zr-MuiDataGrid-virtualScrollerRenderZone"
                style="transform: translate3d(0px, 0px, 0px);"
              >
                <div
                  aria-rowindex="2"
                  aria-selected="false"
                  class="MuiDataGrid-row MuiDataGrid-row--lastVisible"
                  data-id="1"
                  data-rowindex="0"
                  role="row"
                  style="max-height: 52px; min-height: 52px;"
                >
                  <div
                    aria-colindex="1"
                    class="MuiDataGrid-cellCheckbox MuiDataGrid-cell--withRenderer MuiDataGrid-cell MuiDataGrid-cell--textCenter"
                    data-colindex="0"
                    data-field="__check__"
                    role="cell"
                    style="min-width: 50px; max-width: 50px; min-height: 52px; max-height: 52px;"
                    tabindex="-1"
                  >
                    <span
                      api="[object Object]"
                      class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root MuiDataGrid-checkboxInput css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                    >
                      <input
                        aria-label="Select row"
                        class="PrivateSwitchBase-input css-1m9pwf3"
                        data-indeterminate="false"
                        tabindex="-1"
                        type="checkbox"
                      />
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CheckBoxOutlineBlankIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </span>
                  </div>
                  <div
                    aria-colindex="2"
                    class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                    data-colindex="1"
                    data-field="field_1"
                    role="cell"
                    style="min-width: 150px; max-width: 150px; min-height: 52px; max-height: 52px;"
                    tabindex="-1"
                  >
                    <div
                      class="MuiDataGrid-cellContent"
                    >
                      field-1
                    </div>
                  </div>
                  <div
                    aria-colindex="3"
                    class="MuiDataGrid-cell MuiDataGrid-cell--textLeft"
                    data-colindex="2"
                    data-field="field_2"
                    role="cell"
                    style="min-width: 150px; max-width: 150px; min-height: 52px; max-height: 52px;"
                    tabindex="-1"
                  >
                    <div
                      class="MuiDataGrid-cellContent"
                    >
                      field_2
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="Mui-resizeTriggers"
        >
          <div
            class="expand-trigger"
          >
            <div
              style="width: 1px; height: 1px;"
            />
          </div>
          <div
            class="contract-trigger"
          />
        </div>
      </div>
      <div>
        <div
          class="MuiDataGrid-footerContainer css-17jjc08-MuiDataGrid-footerContainer"
        >
          <div />
          <div
            class="MuiTablePagination-root css-rtrcn9-MuiTablePagination-root"
          >
            <div
              class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-11bfvty-MuiToolbar-root-MuiTablePagination-toolbar"
            >
              <div
                class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
              />
              <p
                class="MuiTablePagination-selectLabel css-pdct74-MuiTablePagination-selectLabel"
                id="mui-2"
              >
                Rows per page:
              </p>
              <div
                class="MuiInputBase-root MuiInputBase-colorPrimary css-16c50h-MuiInputBase-root-MuiTablePagination-select"
                variant="standard"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  aria-labelledby="mui-2 mui-1"
                  class="MuiTablePagination-select MuiSelect-select MuiSelect-standard MuiInputBase-input css-194a1fa-MuiSelect-select-MuiInputBase-input"
                  id="mui-1"
                  role="button"
                  tabindex="0"
                >
                  25
                </div>
                <input
                  aria-hidden="true"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  tabindex="-1"
                  value="25"
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiTablePagination-selectIcon MuiSelect-icon MuiSelect-iconStandard css-pqjvzy-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
              </div>
              <p
                class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
              >
                1–1 of 1
              </p>
              <div
                class="MuiTablePagination-actions"
              >
                <button
                  aria-label="Go to previous page"
                  class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                  disabled=""
                  tabindex="-1"
                  title="Go to previous page"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="KeyboardArrowLeftIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                    />
                  </svg>
                </button>
                <button
                  aria-label="Go to next page"
                  class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                  disabled=""
                  tabindex="-1"
                  title="Go to next page"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="KeyboardArrowRightIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
